BUG: Sign of elements in random orthogonal 2D matrices in "ortho_group_gen" class

philipdeboer opened the pull request:
BUG: Fix two bugs with random matrices

* Fix handling of random state
* Fix signs of random orthogonal 2D matrices

pv commented:
I think you need to comment why this is the correct fix.

philipdeboer commented:
Sure, should the note be changed to say:
* Ensure random 2D orthogonal matrices have random determinants
I included a test which checks that both positive and negative determinants are sampled for a variety of dimensions.

pv commented:
Sure, but the concern is that if the algorithm fails for ndim=2, was it correctly implemented?
Failure for ndim>2 might manifest in a different way than just in the sign of the determinant.

pv commented:
I read the paper: is the issue that multiplication by \tilde{H}_1^T missing? If yes, wouldn't it affect results for any `dim`?

philipdeboer commented:
I will look into it!

On 23 November 2017 at 16:04, Pauli Virtanen <notifications@github.com>
wrote:

> I read the paper: is the issue that multiplication by \tilde{H}_1^T
> missing? If yes, wouldn't it affect results for any dim?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/8187#issuecomment-346697277>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJ_rHjHoN4KRzInO_Qf6Escq0n8tQnz8ks5s5d33gaJpZM4QpBmo>
> .
>

pv commented:
Also another thing I notice: I didn't find where the "on line" generated random vectors are discussed in Mezzadri? It seems the idea there was to apply the "improved" QR on a pre-generated fixed matrix with elements from N(0,1).

Did you prove that the "v" vectors ("x" in this algorithm) that appearing during the QR process are distributed as elementwise-N(0,1), similarly as the elements of the initial matrix? (This maybe follows from the orthogonal invariance of GL(N,R), but I didn't think it through now.)

philipdeboer commented:
The key observation that Mezzadri makes is that we should choose a certain sign to avoid large round-off errors. I suspect that this choice restricts the remaining dof sufficiently such that in 2D we always have the same sign for the determinant.

Rather than trying to understand again the details of the various papers I have collected on this subject, I took an experimentalist's approach. We are generating a series of random matrices M, which should have the property that the set {Me} is distributed randomly for some given vector e. This also means that the distributions {Me} and {My} should be the same for distinct vectors e and y (of course, the particular sets {Me} and {My} will not be the same since they are finite samples).

In two dimensions, this means the angles between vectors should be uniformly distributed. For simplicity pick e = [1, 0], a unit vector.
    xs = [ scipy.stats.ortho_group.rvs(dim=2) for i in range(10000) ]
    e = np.array([1,0])
    angles = sorted([ np.arccos(np.dot(e, np.dot(M, e))) for M in xs ])
Now we can plot these angles and see that they are uniformly distributed between 0 and 3.14 as expected.

We can repeat this in 3 dimensions, but now the angles aren't uniformly distributed. That's because we expect fewer points near the poles and more near the equator. Indeed, that's what the distribution shows. Instead of computing an explicit form for this distribution, we can recognize that the distributions of {M . (1,0,0)} and {M . (0,1,0)} are the same. So dot those distributions with another convenient vector and use the two-sided KS test to confirm that the distributions are indeed the same. These projections and KS test generalize to higher dimensions, and are used in the existing test_haar tests. Incidentally I also found an independent python implementation of 3D rotations and confirmed that it gave the same distribution as mine.

The remaining question then is what is going on with the determinants. We can sum the dets of 10000 samples in various dimensions, and show that the sum is close to zero except in 2D. But why does the 2D case pass my "haar" test? The projections don't distinguish reflections, and thus lose the information in the determinant. So the sum of dets is the necessary missing check, and with the submitted patch it is fixed for 2D. I added a check to ensure that there is at least one positive and one negative det in a sample of 10 for each of several small dimensions, which I believe is sufficient to check this.

pv commented:
Multiplication by \tilde{H}_1^T should be a random sign flip in the last 
column. If there's no analytic proof it is not needed except for dim==2, 
it most likely should be included for all dimensions.

The "online" generation of random vectors is discussed in GW Steward, 
SIAM J. Numer. Anal., 17(3), 403 (1980) https://doi.org/10.1137/0717034 
so there's no problem with that.

philipdeboer commented:
So would you just add something like
        H[-1, -1] = (-1)**random_state.randint(2)
before entering the loop, to put in a random sign? I think this is the H1 you were referring to. Is the concern that although the dets are 50/50 in other dimensions, the det could be correlated to the other entries and thus subtly breaking the Haar measure in a way that our tests haven't been precise enough to capture?

ilayn commented:
@philipdeboer I've edited the title a bit such that it is easier for others in terms of maintenance efforts. Please adjust accordingly if it is not a description of the actual problem.

pv commented:
Right --- I mean that the published algorithm should be followed down to the last detail, unless you have mathematical proof of what you are doing is correct. Empirical tests can check for typical failures, but with that approach you cannot obtain full certainty.

philipdeboer commented:
Oh, I agree completely Pauli, and that's one of the reasons I included references in the original submission. I do very much appreciate the effort you took to read the paper and spot the missing detail. My last question regarding H[-1,-1] was simply whether the proposed code agrees with the paper and resolves this issue, or whether I have misunderstood it.

pv commented on scipy/stats/_multivariate.py:
```diff
@@ -3448,16 +3448,17 @@ def rvs(self, dim, size=1, random_state=None):
-        random_state = self._get_random_state(random_state)
-
         H = np.eye(dim)
+        H[-1, -1] = (-1)**random_state.randint(2)
```
This is multiplying by \tilde{H}_1^T from the *wrong side*. See Eq. (7.7) in Mezzadri. You may also want to think if extending the loop index to `dim+1` does the same thing.

pv replied:
Also, in the loop below, please fix
```
D = np.sign(x[0]) if x[0] != 0 else 1
```

pv commented on scipy/stats/_multivariate.py:
```diff
@@ -3514,11 +3515,10 @@ class random_correlation_gen(multi_rv_generic):
+    array([ 1.,       -0.204192,  0.114732,  0.250597],
+          [-0.204192,  1.,       -0.176032, -0.291236],
+          [ 0.114732, -0.176032,  1.,       -0.237777],
+          [ 0.250597, -0.291236, -0.237777,  1.      ]])
```
You need to use more digits, see travis test errors.

philipdeboer commented:
Thanks for your patience; here is an updated patch.

I renamed n-1 -> n, and that helped me understand the issue, which I hope is fixed now for O(N). I've also included Ying's test against a reference implementation, which is passing too. I will also propose these changes to the MDP Toolkit project, since the code originated there.

I modified the SO(N) implementation in a similar way. There the loop does stop at dim-1, but SO(N) has fewer dof than O(N).

However, I did not find the line
D = np.sign(x[0]) if x[0] != 0 else 1
OK, just realized you are proposing this /as/ as fix, since np.sign(z) returns 0 if z == 0. WIll update.

philipdeboer commented:
FYI, I get occasional failures of
TestMatrixNormal.test_frozen_matrix_normal

E               AssertionError: 
E               Items are not equal:
E                ACTUAL: 1.3916022084093**282**e-07
E                DESIRED: 1.3916022084093**305**e-07

The discrepancy is very small though.

pv commented:
thanks merged, looks now better.
